Firestore: Add backup schedule resource

IchordeDionysos opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
Adds support for configuring Firestore backup schedules.

Fixes https://github.com/hashicorp/terraform-provider-google/issues/15790

<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_firestore_backup_schedule`
```

modular-magician commented:
Hello! I am a robot. It looks like you are a: Community Contributor ~Googler~ ~Core Contributor~. Tests will require approval to run.

@SarahFrench, a repository maintainer, has been assigned to review your changes. If you have not received review feedback within 2 business days, please leave a comment on this PR asking them to take a look.

You can help make sure that review is quick by [doing a self-review](https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/) and by [running impacted tests locally](https://googlecloudplatform.github.io/magic-modules/get-started/run-provider-tests/).

IchordeDionysos commented:
@SarahFrench I've seen the build `presubmit-generate-diffs` failed...
Is there anything I could do to fix this? Unfortunately I don't see the logs of the build ...

SarahFrench commented:
https://github.com/GoogleCloudPlatform/magic-modules/blob/3e72aaca0004bc37bdb16fa8de6b6c670d268dd6/mmv1/products/firestore/BackupSchedule.yaml#L76-L82

The build is failing because this field is a `NestedObject` and needs a `properties` field. To let the field be an object with no fields inside ([as is the case for dailyRecurrence](https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases.backupSchedules#DailyRecurrence)) we put `properties: []` in the yaml. Could you please push a commit with that change?

IchordeDionysos commented:
@SarahFrench I pushed the proposed change

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8908-old..auto-pr-8908) ( 4 files changed, 646 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8908-old..auto-pr-8908) ( 4 files changed, 646 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8908-old..auto-pr-8908) ( 1 file changed, 134 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_firestore_backup_schedule` (1 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_firestore_backup_schedule" "primary" {
  collection = # value needed
  database   = # value needed
  weekly_recurrence {
    day = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3026` 
Passed tests `2727` 
Skipped tests: `297` 
Affected tests: `2` 

#### Action taken 
<details> <summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccDataprocClusterIamPolicy|TestAccFirestoreBackupSchedule_firestoreBackupScheduleBasicExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

IchordeDionysos commented on mmv1/products/firestore/BackupSchedule.yaml:
```diff
@@ -0,0 +1,104 @@
+    description: |
+      The Firestore database id. Defaults to `"(default)"`.
+    ignore_read: true
+  - !ruby/object:Api::Type::String
+    name: collection
+    required: true
+    description: |
+      The collection being indexed.
+    ignore_read: true
```
This was a copy & paste error and needs to be removed!

```suggestion
```

IchordeDionysos commented:
@SarahFrench do I read [this report](https://github.com/GoogleCloudPlatform/magic-modules/pull/8908#issuecomment-1714020990) correctly that there needs to be a second example that uses the `weekly_schedule`, and in one of the examples, `database` needs to be explicitly defined (instead of relying on the default)?

`collection` should be removed from the PR!

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccDataprocClusterIamPolicy`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/5ca1d856-d779-45da-9877-9d1ce9ae7dd5/recording/TestAccDataprocClusterIamPolicy.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccFirestoreBackupSchedule_firestoreBackupScheduleBasicExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/5ca1d856-d779-45da-9877-9d1ce9ae7dd5/build-log/recording_build/TestAccFirestoreBackupSchedule_firestoreBackupScheduleBasicExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/5ca1d856-d779-45da-9877-9d1ce9ae7dd5/recording/TestAccFirestoreBackupSchedule_firestoreBackupScheduleBasicExample.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/5ca1d856-d779-45da-9877-9d1ce9ae7dd5/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/5ca1d856-d779-45da-9877-9d1ce9ae7dd5/recording) for each test

IchordeDionysos commented:
@SarahFrench I think I've run into a dead end here ...

This works fine:

```tf
resource "google_firestore_backup_schedule" "daily-backup" {
  retention = "604800s"
  weekly_recurrence {
    day = "MONDAY"
  }
}
```

However, the empty `daily_recurrence` is proving problematic right now ...

```tf
resource "google_firestore_backup_schedule" "daily-backup" {
  retention = "604800s"
  daily_recurrence {}
}
```

When using the above configuration and running `terraform apply` this is the resulting state:
```json
{
  "version": 4,
  "terraform_version": "1.4.4",
  "serial": 21,
  "lineage": "62e0e33f-9a80-4e5b-7d93-9d456620c84e",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "google_firestore_backup_schedule",
      "name": "daily-backup",
      "provider": "provider[\"registry.terraform.io/hashicorp/google\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "daily_recurrence": [],
            "database": "(default)",
            "id": "projects/simpleclub-magic-modules-tests/databases/(default)/backupSchedules/93b9f8ac-e730-48ac-8bff-bf72c43ab22f",
            "name": "projects/simpleclub-magic-modules-tests/databases/(default)/backupSchedules/93b9f8ac-e730-48ac-8bff-bf72c43ab22f",
            "project": "simpleclub-magic-modules-tests",
            "retention": "604800s",
            "timeouts": null,
            "weekly_recurrence": []
          },
          "sensitive_attributes": [],
          "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjoxMjAwMDAwMDAwMDAwLCJkZWxldGUiOjEyMDAwMDAwMDAwMDB9fQ=="
        }
      ]
    }
  ],
  "check_results": null
}
```

Initially, I had problems that `daily_recurrence` was sent as `null` instead of an empty object.
To solve this, I added these two fields to `daily_recurrence`:

```yaml
    custom_expand: 'templates/terraform/custom_expand/empty_object_if_set.go.erb'
    send_empty_value: true
```

This allowed me to apply the changes. 
However, when running `terraform plan` after applying it always wants to recreate the resource, as `daily_recurrence` "got added" (according to Terraform) and this forces replacement:

```tf
> terraform plan
  # google_firestore_backup_schedule.daily-backup must be replaced
-/+ resource "google_firestore_backup_schedule" "daily-backup" {
      ~ id        = "projects/simpleclub-magic-modules-tests/databases/(default)/backupSchedules/93b9f8ac-e730-48ac-8bff-bf72c43ab22f" -> (known after apply)
      ~ name      = "projects/simpleclub-magic-modules-tests/databases/(default)/backupSchedules/93b9f8ac-e730-48ac-8bff-bf72c43ab22f" -> (known after apply)
        # (3 unchanged attributes hidden)

      + daily_recurrence {} # forces replacement
    }
```

As you can see in the state file, there is an empty array for `"daily_recurrence": [],` and one for `"weekly_recurrence": []` making it impossible to distinguish which one of these was set. 
As a human, it's obvious that `daily_recurrence` was used otherwise something would be in the `weekly_recurrence` array.

But the state should look something like that, no?
```
"daily_recurrence": [{}],
```

However, I need help understanding how I can achieve that or how to make the re-planning after applying not result in any changes.

-- 

I know this would have to do something with a custom encoder, decoder, or ... [1] but I need help seeing which one I should use.
I tried a custom encode and custom_expand, but they did not fix the re-planning issue.

@SarahFrench could you point me to the right thing to use here or some hints on how to fix the issue?

[1] https://googlecloudplatform.github.io/magic-modules/develop/custom-code/

SarahFrench commented:
Sorry for the delays reviewing this PR, I've been working to a deadline this past week! I'll take a proper look at this tomorrow

SarahFrench commented:
> Also could you let me know what exactly the error of the test was?

[The error in this message](https://github.com/GoogleCloudPlatform/magic-modules/pull/8908#issuecomment-1714165992) was due to `collection` missing from a configuration used in the acceptance test: `The argument "collection" is required, but no definition was found.` Thanks for removing it from this PR!

---

> @SarahFrench do I read https://github.com/GoogleCloudPlatform/magic-modules/pull/8908#issuecomment-1714020990 correctly that there needs to be a second example that uses the weekly_schedule, and in one of the examples, database needs to be explicitly defined (instead of relying on the default)?

Yes that's correct - ideally acceptance tests will include all of a resource's fields. To address that automated comment you could create a second example file that uses the `weekly_recurrence` field, instead of daily, and set `skip_docs: true` in the `ruby/object:Provider::Terraform::Examples` section for that new example. This would allow the code generator to use that example to make a new acceptance test but avoids both examples being included in the documentation.

SarahFrench commented:
I'm taking a look into https://github.com/GoogleCloudPlatform/magic-modules/pull/8908#issuecomment-1717753498 now, but in the meantime I'll trigger the acceptance tests again

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8908-old..auto-pr-8908) ( 4 files changed, 626 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8908-old..auto-pr-8908) ( 4 files changed, 626 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8908-old..auto-pr-8908) ( 1 file changed, 124 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_firestore_backup_schedule` (1 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_firestore_backup_schedule" "primary" {
  database = # value needed
  weekly_recurrence {
    day = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3043` 
Passed tests `2745` 
Skipped tests: `297` 
Affected tests: `1` 

#### Action taken 
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccFirestoreBackupSchedule_firestoreBackupScheduleBasicExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccFirestoreBackupSchedule_firestoreBackupScheduleBasicExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/ea485773-fb2d-48c3-8fa7-a61e6983168b/build-log/recording_build/TestAccFirestoreBackupSchedule_firestoreBackupScheduleBasicExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/ea485773-fb2d-48c3-8fa7-a61e6983168b/recording/TestAccFirestoreBackupSchedule_firestoreBackupScheduleBasicExample.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/ea485773-fb2d-48c3-8fa7-a61e6983168b/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/ea485773-fb2d-48c3-8fa7-a61e6983168b/recording) for each test

SarahFrench commented on mmv1/products/firestore/BackupSchedule.yaml:
```diff
@@ -0,0 +1,98 @@
+    description: |
+      A server defined name for this index. Format:
+      `projects/{{project}}/databases/{{database}}/backupSchedules/{{backupSchedule}}
+  - !ruby/object:Api::Type::String
+    name: database
+    default_value: '(default)'
+    description: |
+      The Firestore database id. Defaults to `"(default)"`.
+    ignore_read: true
+  - !ruby/object:Api::Type::String
+    name: retention
+    required: true
+    description: |
+      At what relative time in the future, compared to its creation time, the backup should be deleted, e.g. keep backups for 7 days.
+      A duration in seconds with up to nine fractional digits, ending with 's'. Example: "3.5s".
```
I think this field also needs to be removed from the PR? I can't see the `database` field in [the documentation for a backup schedule](https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases.backupSchedules), and currently when I try to run acceptance tests I get this error from the API:


```
  |       {
  |         "description": "Invalid JSON payload received. Unknown name \"database\" at 'backup_schedule': Cannot find field.",
  |         "field": "backup_schedule"
  |       }
```

SarahFrench replied:
Ah I see this is needed to form the URL for the resource - I think that you need to include database as a parameter in the YAML, [similar to this example](https://github.com/GoogleCloudPlatform/magic-modules/blob/c8478f3a43b0f3f03542f5427104760eb5fb482c/mmv1/products/apigateway/ApiConfig.yaml#L93-L107)

IchordeDionysos replied:
Ahh yes thanks, I've updated it.
When looking at various resources I've seen both approaches where it's in the parameters and where it's in the properties, that confused me!
e.g. in the Firestore `Field` is was in the properties :)

SarahFrench commented:
> Initially, I had problems that daily_recurrence was sent as null instead of an empty object.
To solve this, I added these two fields to daily_recurrence:

Please try setting both`send_empty_value: true` and `allow_empty_object: true` for that field - I believe that'll solve the issue you're seeing around that field

IchordeDionysos commented:
@SarahFrench what I've done now:
- Added an example for `weekly_recurrence`.  
  
  I've decided against excluding it from the docs. Having both daily and weekly schedules helps to understand how to use the resource, and allowed me to make it even clearer what values are allowed for `retention` in the two cases. (e.g. the exact value in seconds to use for the maximum allowed retention).  
  
  And given there are very few properties the documentation should not be too long/cluttered with both examples.
- I've moved `database` from the `properties` to the `parameters`.
- I've fixed the `daily_recurrence` according to your suggestion. It now works 😍🚀 
- Tested and fixed the importing of backup schedules
- Tested and fixed the updating of backup schedules
  - Updating the retention is now possible without the replacement of the resource
  - Updating daily and weekly recurrence is possible but forces replacement

If there is anything else I should improve or fix, let me know!

IchordeDionysos commented:
@SarahFrench, I appreciate the detailed review from your side. It helped me understand how the `magic-modules` work and I also really look forward to the 5.0.0 release as it fixes another annoying issue we have right now.

Can you give me some clarity on when you get to review this PR, will it only be after the 5.0.0 release (which already had been pushed back)?

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8908-old..auto-pr-8908) ( 4 files changed, 747 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8908-old..auto-pr-8908) ( 4 files changed, 747 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8908-old..auto-pr-8908) ( 1 file changed, 119 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_firestore_backup_schedule` (2 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_firestore_backup_schedule" "primary" {
  database = # value needed
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3112` 
Passed tests `2790` 
Skipped tests: `314` 
Affected tests: `8` 

#### Action taken 
<details> <summary>Found 8 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccContainerNodePool_withUpgradeSettings|TestAccContainerNodePool_withKubeletConfig|TestAccContainerCluster_withAddons|TestAccDataprocJobIamPolicy|TestAccFirestoreBackupSchedule_firestoreBackupScheduleDailyExample|TestAccFirestoreBackupSchedule_firestoreBackupScheduleWeeklyExample|TestAccDataSourceGoogleServiceAccountJwt|TestAccDataSourceGoogleServiceAccountAccessToken_basic </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccFirestoreBackupSchedule_firestoreBackupScheduleDailyExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording/TestAccFirestoreBackupSchedule_firestoreBackupScheduleDailyExample.log)]
`TestAccFirestoreBackupSchedule_firestoreBackupScheduleWeeklyExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording/TestAccFirestoreBackupSchedule_firestoreBackupScheduleWeeklyExample.log)]
`TestAccDataSourceGoogleServiceAccountJwt`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording/TestAccDataSourceGoogleServiceAccountJwt.log)]
`TestAccDataSourceGoogleServiceAccountAccessToken_basic`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording/TestAccDataSourceGoogleServiceAccountAccessToken_basic.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{red}{\textsf{Tests failed when rerunning REPLAYING mode:}}$ 
 `TestAccDataSourceGoogleServiceAccountJwt`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/build-log/replaying_build_after_recording/TestAccDataSourceGoogleServiceAccountJwt_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/replaying_after_recording/TestAccDataSourceGoogleServiceAccountJwt.log)] 

Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccContainerNodePool_withUpgradeSettings`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/build-log/recording_build/TestAccContainerNodePool_withUpgradeSettings_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording/TestAccContainerNodePool_withUpgradeSettings.log)]
`TestAccContainerNodePool_withKubeletConfig`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/build-log/recording_build/TestAccContainerNodePool_withKubeletConfig_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording/TestAccContainerNodePool_withKubeletConfig.log)]
`TestAccContainerCluster_withAddons`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/build-log/recording_build/TestAccContainerCluster_withAddons_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording/TestAccContainerCluster_withAddons.log)]
`TestAccDataprocJobIamPolicy`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/build-log/recording_build/TestAccDataprocJobIamPolicy_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording/TestAccDataprocJobIamPolicy.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/eb79c81a-4389-4073-b584-874811e9c126/recording) for each test

SarahFrench commented:
@IchordeDionysos don't worry about those tests failing, they're not affected by this PR!

SarahFrench commented on mmv1/products/firestore/BackupSchedule.yaml:
```diff
@@ -0,0 +1,114 @@
+    custom_flatten: templates/terraform/custom_flatten/name_from_self_link.erb
+  - !ruby/object:Api::Type::String
+    name: retention
+    required: true
```
Is `retention` a required field? What happens if it is not included in a request to the API? The API [documentation doesn't note that the field is required](https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases.backupSchedules#BackupSchedule.FIELDS.retention), but that might just be an inaccuracy in the docs.

IchordeDionysos replied:
When trying to create it without it leads to the following error:

```
{
  "error": {
    "code": 400,
    "message": "\u003ceye3 title='/FirestoreAdmin.CreateBackupSchedule, INVALID_ARGUMENT'/\u003e APPLICATION_ERROR;google.firestore.admin.v1/FirestoreAdmin.CreateBackupSchedule;Retention must be specified in the BackupSchedule.;AppErrorCode=3;StartTimeMs=1696353643353;unknown;ResFormat=uncompressed;ServerTimeSec=0.001389596;LogBytes=256;Non-FailFast;EffSecLevel=none;ReqFormat=uncompressed;ReqID=1e6fb03d0ec10188;GlobalID=0;Server=[2002:a05:6699:e29:b0:6c:f434:64f4]:4001",
    "status": "INVALID_ARGUMENT"
  }
}
```
<img width="400" alt="image" src="https://github.com/GoogleCloudPlatform/magic-modules/assets/10195482/cbcb1355-8eb9-4281-a5ef-dbcbd2018a41">

So yes it seems to be required!

SarahFrench replied:
Thanks for looking into this! Sometimes the API documentation explicitly says a field is required and sometimes it doesn't (when it should) so I figured I'd ask

IchordeDionysos commented on mmv1/products/firestore/BackupSchedule.yaml:
```diff
@@ -0,0 +1,114 @@
+      - weekly_recurrence
+    send_empty_value: true
+    allow_empty_object: true
+    immutable: true
```
When trying to patch an existing backup schedule from daily to weekly, it has no effect.

I've created a backup schedule with daily recurrence.
And as you can see in the example below, I've set the schedule to weekly but the schedule is still daily in the response:

<img width="404" alt="image" src="https://github.com/GoogleCloudPlatform/magic-modules/assets/10195482/1427935e-902f-4816-af91-89871bd37947">

---

Regarding the question of what happens with the backups when the backup schedule is deleted, let me say the following:

The docs says that:
> Deleting the source database does not automatically delete related backups.
[source](https://cloud.google.com/firestore/docs/backups#:~:text=Backups%20have%20a%20configurable%20retention%20period%20and%20are%20stored%20until%20the%20retention%20period%20expires%20or%20until%20you%20delete%20the%20backup.%20Deleting%20the%20source%20database%20does%20not%20automatically%20delete%20related%20backups.)

This means backups will persist until reaching the end of the retention period or when [manually deleting the backup](https://cloud.google.com/firestore/docs/reference/rest/v1/projects.locations.backups/delete).

While the docs do not mention what happens when only the backup schedule is deleted, I strongly think that the same applies to backup schedules.
Deleting a schedule does NOT delete the backups created using this backup schedule.

IchordeDionysos replied:
I can verify the exact behavior tomorrow though when I finally have a backup and can delete the schedule again.

SarahFrench replied:
Brilliant, thank you!

If the update truly doesn't work, despite there being no API errors, then forcing recreation of the resource in Terraform makes sense. But we definitely want to avoid any unintended consequences of that.

If the update does work but has an 'eventual consistency' issue then there might be different approaches to the problem. If you wait a few minutes and then GET the backup schedule you updated, do the new recurrence settings show up?

IchordeDionysos replied:
### Regarding changing the recurrence:

Even one day later, the recurrence is still daily, so there is no eventual consistency thing going on here.
It seems like updating the recurrence without deleting and re-creating is not supported.

### Regarding the backups:

These are the requests I've done:

`GET /projects/<project>/locations/-/backups`
![image](https://github.com/GoogleCloudPlatform/magic-modules/assets/10195482/b2131049-2557-472f-abaa-86e93fd5f5b6)

`GET /projects/<project>/databases/<database>/backupSchedules`
![image](https://github.com/GoogleCloudPlatform/magic-modules/assets/10195482/d16c7a5e-8df2-4899-86dc-42fb6001e91f)

`DELETE /projects/<project>/databases/<database>/backupSchedules/<schedule>`
![image](https://github.com/GoogleCloudPlatform/magic-modules/assets/10195482/2ecf6afc-aa1c-40b4-93e0-e7749dc8c4fa)

`GET /projects/<project>/databases/<database>/backupSchedules`
![image](https://github.com/GoogleCloudPlatform/magic-modules/assets/10195482/b530ce34-db4a-47a2-a4f1-f75845953fbb)

`GET /projects/<project>/locations/-/backups`
![image](https://github.com/GoogleCloudPlatform/magic-modules/assets/10195482/591cd7c3-7727-42da-93f4-c701259f0860)

So even after deleting the backup schedule the backup is still there!
Meaning deletion of the schedule should have no issue or should not need any special consideration.

We may want to expose the backups as data source in the future potentially though.

IchordeDionysos replied:
@SarahFrench does that answer your questions?

rwhogg replied:
(Firestore developer speaking) Changing an existing backup schedule from weekly to daily (or vice-versa) is not supported. If you want to make such a change, you can delete the weekly schedule and then create a new one with daily recurrence.

IchordeDionysos replied:
Perfect then the current behavior should be fine.

Essentially if the recurrence changes the resource has to be deleted and re-created.

-- 

@rwhogg this behavior should not be a problem as previous backups done with a different schedule/recurrence are kept even though the backup schedule has been deleted, no?

SarahFrench replied:
I'm going to mark this thread as resolved, as we now know that the recurrence definitely can't be updated and from your [manual experiments](https://github.com/GoogleCloudPlatform/magic-modules/pull/8908#discussion_r1345910631) there doesn't seem to be a risk of accidentally causing users to lost backups.

rwhogg commented on mmv1/products/firestore/BackupSchedule.yaml:
```diff
@@ -0,0 +1,114 @@
+    `"FIRESTORE_NATIVE"` and `location_id` set to your chosen location.
+    If you wish to use App Engine, you may instead create a
+    `google_app_engine_application` resource with `database_type` set to
+    `"CLOUD_FIRESTORE"`. Your Firestore location will be the same as
```
Can we remove the references to FIRESTORE_NATIVE and CLOUD_FIRESTORE here? Backup schedules work with Firestore in Datastore Mode as well.

IchordeDionysos replied:
@rwhogg updated it :)

rwhogg commented on mmv1/products/firestore/BackupSchedule.yaml:
```diff
@@ -0,0 +1,114 @@
+    name: name
+    output: true
+    description: |
+      A server defined name for this index. Format:
```
Can we put something like:

`The unique backup schedule identifier across all locations and databases for the given project.`

instead?

IchordeDionysos replied:
@rwhogg updated it

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8908-old..auto-pr-8908) ( 4 files changed, 765 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8908-old..auto-pr-8908) ( 4 files changed, 765 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8908-old..auto-pr-8908) ( 1 file changed, 119 insertions(+))

modular-magician commented:
#### Tests analytics 
Total tests: `3127` 
Passed tests `2804` 
Skipped tests: `318` 
Affected tests: `5` 

#### Action taken 
<details> <summary>Found 5 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccAlloydbUser_updatePassword_BuiltIn|TestAccAlloydbUser_updateRoles_BuiltIn|TestAccAlloydbUser_updateRoles_IAM|TestAccAlloydbUser_alloydbUserBuiltinExample|TestAccAlloydbUser_alloydbUserIamExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccAlloydbUser_updatePassword_BuiltIn`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/build-log/recording_build/TestAccAlloydbUser_updatePassword_BuiltIn_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/recording/TestAccAlloydbUser_updatePassword_BuiltIn.log)]
`TestAccAlloydbUser_updateRoles_BuiltIn`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/build-log/recording_build/TestAccAlloydbUser_updateRoles_BuiltIn_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/recording/TestAccAlloydbUser_updateRoles_BuiltIn.log)]
`TestAccAlloydbUser_updateRoles_IAM`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/build-log/recording_build/TestAccAlloydbUser_updateRoles_IAM_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/recording/TestAccAlloydbUser_updateRoles_IAM.log)]
`TestAccAlloydbUser_alloydbUserBuiltinExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/build-log/recording_build/TestAccAlloydbUser_alloydbUserBuiltinExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/recording/TestAccAlloydbUser_alloydbUserBuiltinExample.log)]
`TestAccAlloydbUser_alloydbUserIamExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/build-log/recording_build/TestAccAlloydbUser_alloydbUserIamExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/recording/TestAccAlloydbUser_alloydbUserIamExample.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8908/artifacts/8f9646b8-15e8-48e3-9c01-1120e1f93870/recording) for each test

SarahFrench reviewed (approved):
Approving after review - the failing VCR tests are unrelated to this PR!
